src/factors/Factors.jl

```diff
@@ -165,3 +188,63 @@ function DiscreteOrderedUniformDist(
 
     return DiscreteNonParametric(options, fill(1.0 / n_opts, n_opts))
 end
+
+struct CategoricalDistribution{T}
+    categories::Vector{T}
+    distribution::Categorical
+end
+
+"""
+    CategoricalDistribution(categories::Vector{T}, weights::Vector{Float64})::CategoricalDistribution{T} where T
+    CategoricalDistribution(categories::Vector{T})::CategoricalDistribution{T} where T
+    CategoricalDistribution(categories...)::CategoricalDistribution
+
+Construct a categorical variable. Default to a uniform cateforical variable if the 
+probability weightings are not provided.
+"""
+function CategoricalDistribution(categories::Vector{T}, weights::Vector{Float64})::CategoricalDistribution{T} where T
+    if length(unique(categories)) != length(categories)
+        throw(ArgumentError("Categories in a categorical variable must be unique."))
+    end
+    if length(categories) != length(weights)
+        msg = "Length of categories and weightings do not match."
+        msg *= " Got $(length(categories)) and $(length(weights))."
+        throw(ArgumentError(msg))
+    end
+    if !(sum(weights) ≈ 1) && all(weights .>= 0.0)
+        throw(ArgumentError("Weights must sum to one."))
+    end
+    return CategoricalDistribution(
+        categories,
+        Categorical(weights))
+end
+function CategoricalDistribution(categories::Vector{T})::CategoricalDistribution{T} where T
+    n_categories::Int64 = length(categories)
+    return CategoricalDistribution(categories, fill(1/n_categories, n_categories))
+end
+function CategoricalDistribution(categories...)::CategoricalDistribution
+    cats = collect(categories)
+    return CategoricalDistribution(cats)
+end
+
+function Distributions.quantile(dist::CategoricalDistribution{T}, q::Real)::T where T
+    underlying_idx::Int64 = Distributions.quantile(dist.distribution, q)
+    return dist.categories[underlying_idx]
+end
+
+"""
+    lower_bound(dist::CategoricalDistribution{T})::T where T <: Real
```

Comment:
```suggestion
    lower_bound(dist::CategoricalDistribution{T})::T where {T<:Real}
```

Do you need to support integers too? If not, I suggest `where {T<:AbstractFloat}` (more constrained than `<:Real`)